### ISSUE TITLE
False positive for `implicit-str-concat` when some but not all strings are raw

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#6663]
### Bug description

I don't think `implicit-str-concat` is a useful warning when the reason for breaking and restarting the string is to start or stop treating a portion of a string as a raw string. (As I understand it, we're trying to someone forgetting a comma between arguments.)

For instance, this string needs to escape `\o` but not `\n`:

```python
import unittest

class MyTest(unittest.TestCase):
    def testColors(self):
        self.assertEqual(
            self.string_output(),
            r'\override Stem.color = "darkgreen"' '\n'
        )
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:7: [W1404(implicit-str-concat), ] Implicit string concatenation found in call

---------------------------------------------------------------------
Your code has been rated at 7.50/10 (previous run: -40.00/10, +47.50)
```

### Expected behavior

no msg

### Pylint version

```shell
pylint 2.14.0-b1
astroid 2.12.0-dev0
Python 3.10.1 (v3.10.1:2cd268a3a9, Dec  6 2021, 14:28:59) [Clang 13.0.0 (clang-1300.0.29.3)]
```

**Comment by Pierre-Sassoulas** [comment]
It make sense. I think the problem existed before we extended this check to ``call`` so it was already released just not as likely to be reported.

**Comment by jacobtylerwalls** [comment]
When working on this, we could check for f-strings mixed with normal strings also. You might terminate an f-string so that it's easier to use `{` without escaping.

**Comment by DanielNoord** [comment]
This is hard to solve as `ast` doesn't store "rawness". See https://github.com/PyCQA/astroid/issues/1156.

**Comment by jhbuhrman** [comment]
Shouldn't we be heading for another direction in solving this? I believe the main _intention_ of this particular check is to prevent accidental concatenation while adding members to a list or tuple, or perhaps indeed also when passing parameters in a function call. For example:
```python
COLORS = [
    "red"
    "yellow",
    "blue",
]
```
(note the missing comma after `"red"`).

I frequently use implicit string concatenation when breaking up too long strings (to prevent or fix `line-too-long` warnings). In doing so, the string parts that I want to be concatenated end up between parentheses, ref. the following example:
```python
SENTENCES = [
    "This is a rather short sentence.",
    (
        "This is quite a long sentence, having all kinds of additional"
        " subordinate clauses, however."
    ),
    "This is another short sentence here.",
]
```
(shown as how `black` would reformat this).

So my plea is to *only* warn for string concatenation when the parts are _not_ collected together in parentheses.

For now, I have to generally disable this check, to my regrets, since the `COLORS` example is a good one to be warned about.

**Comment by DanielNoord** [comment]
I think you should open a separate issue for this. This is mainly about concatenating different types.

**Comment by jhbuhrman** [comment]
> I think you should open a separate issue for this. This is mainly about concatenating different types.

Yeah you're right, I discovered that my issue is substantially different (is more about that I've activated `check-str-concat-over-line-jumps`). Thanks.

(Although my remark could provide some sort of work-around for this issue as well? ...)

**Comment by Pierre-Sassoulas** [comment]
We favor false negatives, so this is a trade off I think we should make. There's two problem in ``r"\five" "\six", ``, the implicit string concat and the missing ``r`` before six.

**Comment by jhbuhrman** [comment]
Yes, I should have come up with a better example, but I think we can imagine examples where it is complete logical to enter one value within a list of strings as a raw string, and another as a plain string, and perhaps a third as a f-string. A missing comma doesn't automatically imply that the concatenation is intentional.

\[\[\[

> Although _technically_ there's nothing wrong (yet) with unrecognized escape sequences (as in `"\six"`) I can imagine that it is worthwhile to have a check that doesn't allow for it (especially now I'm reading in https://docs.python.org/3.10/reference/lexical_analysis.html?highlight=escape#escape-sequences that these unrecognized escape sequences are triggering a `DeprecationWarning` since py36).

\]\]\]

Anyhow, that's another issue. As stated [earlier](#issuecomment-1342620613), given the example at the [top of this issue](#issue-1244250500), I would propose to "solve" it by allowing implicit string concatenation if and only if you surround it with a set of parentheses (as a signal to your linter that the concatenation _is_ intentional), as in:
```python
import unittest

class MyTest(unittest.TestCase):
    def testColors(self):
        self.assertEqual(
            self.string_output(),
            (r'\override Stem.color = "darkgreen"' '\n')
        )
```
(ref. https://github.com/PyCQA/pylint/issues/7929)

If I'm not mistaken, the `black` formatter leaves the (sort of redundant) set parentheses alone...
